logger.go

```diff
@@ -0,0 +1,234 @@
+		// Put the buffer back into the pool
+		bufferPool.Put(buf)
+	})
+}
+
+// writerProxy is a proxy around an http.ResponseWriter that allows you to hook
+// into various parts of the response process.
+type writerProxy interface {
+	http.ResponseWriter
+	// Status returns the HTTP status of the request, or 0 if one has not
+	// yet been sent.
+	Status() int
+	// BytesWritten returns the total number of bytes sent to the client.
+	BytesWritten() int
+	// Tee causes the response body to be written to the given io.Writer in
+	// addition to proxying the writes through. Only one io.Writer can be
+	// tee'd to at once: setting a second one will overwrite the first.
+	// Writes will be sent to the proxy before being written to this
+	// io.Writer. It is illegal for the tee'd writer to be modified
+	// concurrently with writes.
+	Tee(io.Writer)
+	// Unwrap returns the original proxied target.
+	Unwrap() http.ResponseWriter
+}
+
+// WrapWriter wraps an http.ResponseWriter, returning a proxy that allows you to
+// hook into various parts of the response process.
+func wrapWriter(resp http.ResponseWriter) writerProxy {
+	_, closeNotifier := resp.(http.CloseNotifier)
+	_, flusher := resp.(http.Flusher)
+	_, hiJacker := resp.(http.Hijacker)
+	_, readerFrom := resp.(io.ReaderFrom)
+
+	basicWriter := basicWriter{ResponseWriter: resp}
+	if closeNotifier && flusher && hiJacker && readerFrom {
+		return &fancyWriter{basicWriter}
+	}
+	if flusher {
+		return &flushWriter{basicWriter}
+	}
+	return &basicWriter
+}
+
+// basicWriter wraps a http.ResponseWriter that implements the minimal
+// http.ResponseWriter interface.
+type basicWriter struct {
+	http.ResponseWriter
+	wroteHeader bool
+	code        int
+	bytes       int
+	tee         io.Writer
+}
+
+func (self *basicWriter) WriteHeader(code int) {
+	if !self.wroteHeader {
+		self.code = code
+		self.wroteHeader = true
+		self.ResponseWriter.WriteHeader(code)
+	}
+}
+func (self *basicWriter) Write(buf []byte) (int, error) {
+	self.WriteHeader(http.StatusOK)
+	n, err := self.ResponseWriter.Write(buf)
+	if self.tee != nil {
+		_, err2 := self.tee.Write(buf[:n])
+		// Prefer errors generated by the proxied writer.
+		if err == nil {
+			err = err2
+		}
+	}
+	self.bytes += n
+	return n, err
+}
+func (self *basicWriter) maybeWriteHeader() {
+	if !self.wroteHeader {
+		self.WriteHeader(http.StatusOK)
+	}
+}
+func (self *basicWriter) Status() int {
+	return self.code
+}
+func (self *basicWriter) BytesWritten() int {
+	return self.bytes
+}
+func (self *basicWriter) Tee(writer io.Writer) {
+	self.tee = writer
+}
+func (self *basicWriter) Unwrap() http.ResponseWriter {
+	return self.ResponseWriter
+}
+
+// fancyWriter is a writer that additionally satisfies http.CloseNotifier,
+// http.Flusher, http.Hijacker, and io.ReaderFrom. It exists for the common case
+// of wrapping the http.ResponseWriter that package http gives you, in order to
+// make the proxied object support the full method set of the proxied object.
```

Comment:
@obukhov-sergey After a bit of experimentation, looks like you are correct, we can use encapsulation.

The following appears to work great! Thanks!
```go
// WrapWriter wraps an http.ResponseWriter, returning a proxy that allows you to
// hook into various parts of the response process.
func wrapWriter(resp http.ResponseWriter) writerProxy {
	closeNotifier := resp.(http.CloseNotifier)
	flusher := resp.(http.Flusher)
	hiJacker := resp.(http.Hijacker)
	readerFrom := resp.(io.ReaderFrom)

	return &WrappedWriter{
		ReaderFrom:     readerFrom,
		Hijacker:       hiJacker,
		Flusher:        flusher,
		CloseNotifier:  closeNotifier,
		ResponseWriter: resp}
}

// Wraps a http.ResponseWriter includes all possible interfaces
type WrappedWriter struct {
	io.ReaderFrom
	http.Hijacker
	http.Flusher
	http.CloseNotifier
	http.ResponseWriter
	wroteHeader bool
	code        int
	bytes       int
	tee         io.Writer
}


var _ http.CloseNotifier = &WrappedWriter{}
var _ http.Flusher = &WrappedWriter{}
var _ http.Hijacker = &WrappedWriter{}
var _ io.ReaderFrom = &WrappedWriter{}
var _ http.Flusher = &WrappedWriter{}
```